src/backend/cdb/cdbmutate.c

```diff
@@ -753,15 +770,9 @@ apply_motion_mutator(Node *node, ApplyMotionState *context)
 	/* An expression node might have subtrees containing plans to be mutated. */
 	if (!is_plan_node(node))
 	{
-		/*
-		 * The containMotionNodes flag keeps track of whether there are any
-		 * Motion nodes, ignoring any in InitPlans. So if we recurse into an
-		 * InitPlan, save and restore the flag.
-		 */
 		if (IsA(node, SubPlan) &&((SubPlan *) node)->is_initplan)
 		{
 			bool		found;
-			bool		saveContainMotionNodes = context->containMotionNodes;
```

Comment:
Why don't we save the number of motions for init plans as it was done before?
And because of this, we get unnecessary(?) Explicit Redistribute Motion in such query:
```sql
explain with cte as (
    select count(*) as c from t, d where t.b = d.b
) update d set b = 10 where a = (select * from cte);
```
```
                                                      QUERY PLAN                                                       
-----------------------------------------------------------------------------------------------------------------------
 Update on d  (cost=1618.58..1622.83 rows=1 width=14)
   InitPlan 1 (returns $0)  (slice4)
     ->  Aggregate  (cost=1618.56..1618.57 rows=1 width=8)
           ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=1618.50..1618.55 rows=1 width=8)
                 ->  Aggregate  (cost=1618.50..1618.51 rows=1 width=8)
                       ->  Hash Join  (cost=11.75..1596.97 rows=2870 width=0)
                             Hash Cond: (t.b = d_1.b)
                             ->  Seq Scan on t  (cost=0.00..961.00 rows=28700 width=4)
                             ->  Hash  (cost=8.00..8.00 rows=100 width=4)
                                   ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..8.00 rows=100 width=4)
                                         ->  Seq Scan on d d_1  (cost=0.00..4.00 rows=34 width=4)
   ->  Explicit Redistribute Motion 3:3  (slice3; segments: 3)  (cost=0.00..4.25 rows=1 width=14)
         ->  Seq Scan on d  (cost=0.00..4.25 rows=1 width=14)
               Filter: (a = $0)
 Optimizer: Postgres query optimizer

```

Reply:
This Explicit Redistribute Motion is definitely unnecessary. Earlier, we ignored motions in InitPlans, since they are executed separately. I brought that logic back